src/core/SampleBuffer.cpp

```diff
@@ -869,122 +431,109 @@ bool SampleBuffer::play(
 	return true;
 }
 
+f_cnt_t SampleBuffer::advance(f_cnt_t playFrame, f_cnt_t frames,  LoopMode loopMode, Sample* state)
+{
+	switch (loopMode)
+	{
+		case LoopMode::LoopOff:
+			return playFrame + frames;
+		case LoopMode::LoopOn:
+			return frames + getLoopedIndex(playFrame, m_playMarkers.loopStartFrame, m_playMarkers.loopEndFrame);
+		case LoopMode::LoopPingPong:
+		{
+			f_cnt_t left = frames;
+			if (state->isBackwards())
+			{
+				playFrame -= frames;
+				if (playFrame < m_playMarkers.loopStartFrame)
+				{
+					left -= m_playMarkers.loopStartFrame - playFrame;
+					playFrame = m_playMarkers.loopStartFrame;
+				}
+				else
+				{
+					left = 0;
+				}
+			}
 
+			return left + getPingPongIndex(playFrame, m_playMarkers.loopStartFrame, m_playMarkers.loopEndFrame);
+		}
+		default:
+			return playFrame;
+	}
+}
 
-
-sampleFrame * SampleBuffer::getSampleFragment(
+std::vector<sampleFrame> SampleBuffer::getSampleFragment(
 	f_cnt_t index,
 	f_cnt_t frames,
 	LoopMode loopMode,
-	sampleFrame * * tmp,
 	bool * backwards,
 	f_cnt_t loopStart,
 	f_cnt_t loopEnd,
 	f_cnt_t end
 ) const
 {
-	if (loopMode == LoopOff)
+	auto out = std::vector<sampleFrame>(frames);
+	if ((loopMode == LoopMode::LoopOff && index + frames <= end) ||
+		(loopMode == LoopMode::LoopOn && index + frames <= loopEnd) ||
+		(loopMode == LoopMode::LoopPingPong && !*backwards && index + frames < loopEnd))
 	{
-		if (index + frames <= end)
-		{
-			return m_data + index;
-		}
+		std::copy_n(m_data.begin() + index, frames, out.begin());
+		return out;
 	}
-	else if (loopMode == LoopOn)
-	{
-		if (index + frames <= loopEnd)
-		{
-			return m_data + index;
-		}
-	}
-	else
+
+	if (loopMode == LoopMode::LoopOff)
 	{
-		if (!*backwards && index + frames < loopEnd)
-		{
-			return m_data + index;
-		}
+		const auto available = end - index;
+		std::copy_n(m_data.begin() + index, available, out.begin());
+		std::fill_n(out.begin() + available, frames - available, sampleFrame{});
+		return out;
 	}
 
-	*tmp = MM_ALLOC<sampleFrame>( frames);
+	auto numFramesCopied = *backwards ?
+		std::min(frames, index - loopStart) :
+		std::min(frames, loopEnd - index);
 
-	if (loopMode == LoopOff)
-	{
-		f_cnt_t available = end - index;
-		memcpy(*tmp, m_data + index, available * BYTES_PER_FRAME);
-		memset(*tmp + available, 0, (frames - available) * BYTES_PER_FRAME);
-	}
-	else if (loopMode == LoopOn)
+	auto updateBackwards = [&]()
```

Comment:
I think this version of `updateBackwards` may not work properly if `loopEnd - loopStart < frames`. I also think you can start by initializing `numFramesCopied` with 0 to reduce duplications.

Reply:
Nevermind, I misunderstood the logic at the first look.

Reply:
> Nevermind, I misunderstood the logic at the first look.

Even though I marked this as resolved, I have been thinking about this recently and decided to split the functionality of this monstrous ``getSampleFragment`` function into more modular ones: ``playSampleRange``, ``playSampleRangeLoop`` ``playSampleRangePingPong`` and ``playSampleRangeBackwards``. Hopefully this makes the code easier to follow, and after finishing my implementation of them there should not be any issues relating to distortion of the signal on playback and the like. I also added the self-explanatory ``resampleSampleRange`` and ``amplifySampleRange`` functions to split things up a bit more.

Feel free to review this a second time if you so choose, each function only spans a few lines so it should be a lot easier to follow.